File: src/main/java/com/github/joelws/release/tracker/service/release/operation/ReadReleaseServiceOperation.java
package com.github.joelws.release.tracker.service.release.operation;

import com.github.joelws.release.tracker.conversion.ReleaseToReleaseDtoAdapter;
import com.github.joelws.release.tracker.dto.release.ReleaseDto;
import com.github.joelws.release.tracker.entity.release.Release;
import com.github.joelws.release.tracker.handler.JsonResponse;
import com.github.joelws.release.tracker.service.ServiceHelper;
import com.github.joelws.release.tracker.service.ServiceOperation;
import com.github.joelws.release.tracker.service.release.execution.ReadReleaseServiceExecution;

import javax.ws.rs.core.Response;

public class ReadReleaseServiceOperation extends ServiceOperation<String> {
    private final ServiceHelper helper;

    private final ReadReleaseServiceExecution readReleaseServiceExecution;

    public ReadReleaseServiceOperation(final ServiceHelper helper,
                                       final ReadReleaseServiceExecution readReleaseServiceExecution) {
        this.helper = helper;
        this.readReleaseServiceExecution = readReleaseServiceExecution;
    }

    @Override
    protected Response delegate(String identifier) {
        final Release result = readReleaseServiceExecution.execute(identifier);

        Response response = null;

        if (result != null) {
            final ReleaseToReleaseDtoAdapter adapter = helper.getFactory().getImpl(ReleaseToReleaseDtoAdapter.class);
            final ReleaseDto adaptedResult = adapter.adapt(result);
            response = helper.getRestResponseBuilder().build(200, adaptedResult);
        } else {
            response = helper.getRestResponseBuilder().build(new JsonResponse(404, "Release doesn't exist"));
        }
        return response;
    }
}

File: src/main/kotlin/com/github/joelws/release/tracker/service/release/operation/ReadReleaseServiceOperation.kt
package com.github.joelws.release.tracker.service.release.operation

import com.github.joelws.release.tracker.conversion.ReleaseAdapter
import com.github.joelws.release.tracker.entity.release.Release
import com.github.joelws.release.tracker.interfaces.Adapter
import com.github.joelws.release.tracker.model.release.ReleaseModel
import com.github.joelws.release.tracker.response.RestResponse.NotFound
import com.github.joelws.release.tracker.response.RestResponse.SuccessWithEntity
import com.github.joelws.release.tracker.response.build
import com.github.joelws.release.tracker.service.ServiceExecution
import com.github.joelws.release.tracker.service.ServiceHelper
import com.github.joelws.release.tracker.service.ServiceOperation
import javax.ws.rs.core.Response

open class ReadReleaseServiceOperation(private val helper: ServiceHelper,
                                       private val readReleaseServiceExecution: ServiceExecution<String?, Release?>) : ServiceOperation<String> {

    override fun delegate(param: String?): Response {

        val result = readReleaseServiceExecution.execute(param)

        return if (result != null) {

            @Suppress("UNCHECKED_CAST")
            val releaseAdapter: Adapter<Release, ReleaseModel> = helper.adapterFactory.getAdapter(ReleaseAdapter::class.java) as Adapter<Release, ReleaseModel>

            val adaptedResult = releaseAdapter.adapt(result)

            SuccessWithEntity(adaptedResult).build()

        } else {
            NotFound("Release doesn't exist").build()
        }
    }
}
